### PR TITLE
only set changed attributes in mergeAttrs (#2984)

### LINKED ISSUE #2984: Minimum on <input type="date"> causes browser state to be lost while typing date on Chrome
### Environment

* Elixir version (elixir -v): 1.15.7
* Phoenix version (mix deps): 1.7.10
* Phoenix LiveView version (mix deps): 0.20.3
* Operating system: macOS 14.1.2
* Browsers you attempted to reproduce this bug on (the more the merrier): Chrome (does not repro on Firefox, Safari)
* Does the problem persist after removing "assets/node_modules" and trying again? Yes/no: Yes

### Actual behavior

When using an `<input type="date">` with the `min` attribute set, typing in a full date (month, day, year) causes Chrome to reset some internal state

Example live view that reproduces this:
```elixir
defmodule LiveDateTestWeb.TestLive do
  use LiveDateTestWeb, :live_view
  def mount(_params, _session, socket) do
    {:ok, assign(socket, data: %{"date" => ""})}
  end
  def render(assigns) do
    ~H"""
    <.form :let={f} for={@data}>
      <%= Phoenix.HTML.Form.date_input(f, :date, phx_change: "date", min: "2023-01-04") %>
    </.form>
    """
  end
  def handle_event("date", %{"date" => date}, socket) do
    {:noreply, assign(socket, data: %{"date" => date})}
  end
end
```

Steps to reproduce:
1. Visit the live view in Chrome
2. Select the month field of the date input
3. Type in a month and day
4. Attempt to type in a year
5. Observe that typing in the second character of the year fails (e.g., typing 2345 for the year results in the year field reading 0005)

In the following screen recording, my keystrokes were `11112345`:

https://github.com/phoenixframework/phoenix_live_view/assets/7091588/ad9478d4-3f61-4d1e-8509-ac74aaf63452

### Expected behavior

Typing in a date does not cause the browser to lose any state and the user can finish typing the year successfully.

---

The underlying cause of the issue is twofold:

1. When merging the updates into the DOM, LiveView always calls `setAttribute` regardless of whether the value actually changes (`min` does not): https://github.com/phoenixframework/phoenix_live_view/blob/dd89eea5b70d0c89a3a7148ddab093a0de6393d9/assets/js/phoenix_live_view/dom.js#L363
2. When the `min`/`max` attributes are set (regardless of whether the value changed), Chrome always destroys and recreates the shadow DOM of the input, which effectively resets the type ahead buffer. (The old element had no value and a type ahead buffer containing `2`, and the new element has a value of `2` and an empty type ahead buffer. So, when the next character (`3`) is typed, it's the only thing in the buffer and replaces the actual value of `2`).
    1. https://github.com/chromium/chromium/blob/4de1cec1246698932cde06c8d628130b4541a93c/third_party/blink/renderer/core/html/forms/multiple_fields_temporal_input_type_view.cc#L526-L528
    2. https://github.com/chromium/chromium/blob/4de1cec1246698932cde06c8d628130b4541a93c/third_party/blink/renderer/core/html/forms/date_time_edit_element.cc#L780-L810

### PR DESCRIPTION
**Pull request by SteffenDE**
Fixes #2984.

Not sure if this is a hot path though, so the extra checking might not be what we want?

### CODE REVIEW

**Comment by SteffenDE** [comment]
Single file example to reproduce the fix:

```elixir
Application.put_env(:sample, Example.Endpoint,
  http: [ip: {127, 0, 0, 1}, port: 5001],
  server: true,
  live_view: [signing_salt: "aaaaaaaa"],
  secret_key_base: String.duplicate("a", 64)
)

Mix.install([
  {:plug_cowboy, "~> 2.5"},
  {:jason, "~> 1.0"},
  {:phoenix, "~> 1.7.10", override: true},
  {:phoenix_live_view, github: "phoenixframework/phoenix_live_view", branch: "main"}
])

defmodule Example.ErrorView do
  def render(template, _), do: Phoenix.Controller.status_message_from_template(template)
end

defmodule Example.HomeLive do
  use Phoenix.LiveView, layout: {__MODULE__, :live}

 def mount(_params, _session, socket) do
    socket
    |> then(&{:ok, &1})
  end

  def render("live.html", assigns) do
    ~H"""
    <script src="https://cdn.jsdelivr.net/npm/phoenix@1.7.10/priv/static/phoenix.min.js"></script>
    <script src="https://cdn.jsdelivr.net/gh/SteffenDE/phoenix_live_view@only_merge_changes_assets/priv/static/phoenix_live_view.js"></script>
    <script>
      let liveSocket = new window.LiveView.LiveSocket("/live", window.Phoenix.Socket, {
        hooks: {
          FakeHook: {
            mounted() {}
          }
        }
      })
      liveSocket.connect()
    </script>
    <style>
      * { font-size: 1.1em; }
    </style>
    <%= @inner_content %>
    """
  end

  def render(assigns) do
    ~H"""
    <%= live_render(@socket, Example.TestLive, id: "child") %>
    """
  end
end

defmodule Example.TestLive do
  use Phoenix.LiveView
  def mount(_params, _session, socket) do
    {:ok, assign(socket, data: %{"date" => ""})}
  end
  def render(assigns) do
    ~H"""
    <.form :let={f} for={@data}>
      <%= Phoenix.HTML.Form.date_input(f, :date, phx_change: "date", min: "2023-01-04") %>
    </.form>
    """
  end
  def handle_event("date", %{"date" => date}, socket) do
    {:noreply, assign(socket, data: %{"date" => date})}
  end
end

defmodule Example.LiveComponent do
  use Phoenix.LiveComponent

  def render(assigns) do
    ~H"""
    <li id={@id}>
      <%= @item.name %>
    </li>
    """
  end
end

defmodule Example.Router do
  use Phoenix.Router
  import Phoenix.LiveView.Router

  pipeline :browser do
    plug(:accepts, ["html"])
  end

  scope "/", Example do
    pipe_through(:browser)

    live("/", HomeLive, :index)
  end
end

defmodule Example.Endpoint do
  use Phoenix.Endpoint, otp_app: :sample
  socket("/live", Phoenix.LiveView.Socket)
  plug(Example.Router)
end

{:ok, _} = Supervisor.start_link([Example.Endpoint], strategy: :one_for_one)
Process.sleep(:infinity)
```

File: assets/js/phoenix_live_view/dom.js [hunk 1]
```diff
@@ -360,7 +360,11 @@ let DOM = {
     let sourceAttrs = source.attributes
     for(let i = sourceAttrs.length - 1; i >= 0; i--){
       let name = sourceAttrs[i].name
-      if(exclude.indexOf(name) < 0){ target.setAttribute(name, source.getAttribute(name)) }
+      if(exclude.indexOf(name) < 0){
+        if(target.getAttribute(name) !== source.getAttribute(name)){
+          target.setAttribute(name, source.getAttribute(name))
+        }
+      }
```
Comment (vanderhoop) [inline assets/js/phoenix_live_view/dom.js:367]: Regarding performance, the additional check has a small cost, but the existing code could be optimized, as `exclude.indexOf` is being called within a loop of the `sourceAttrs. 

Converting `exclude` to a `Set` for those checks would [in most cases](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set#performance) avoid  nested iteration under the hood.

It's existing code, but a small change to make, and a good pattern to follow.

**Comment by josevalim** [comment]
Performance wise, it may be worth having two functions: mergeAttrs and mergeOnlyChangedAttrs, and we use the slow version only on inputs, instead of everywhere else `mergeAttrs` is called today. Don't listen to me though, @chrismccord will know better.

**Comment by shadowfacts** [comment]
I'll defer to @chrismccord as well, but it's not entirely clear to me why the LiveView code needs to set _any_ attributes—I was under the impression that all of that is handled by morphdom (which does seem to check whether the actual value has changed).

**Comment by chrismccord** [comment]
❤️❤️❤️🐥🔥

**Pull request merged**